### PR TITLE
docs: add EnableDefaultDeny documentation

### PR DESCRIPTION
**Pull request by squeed**
This adds more complication to the logic for determining an endpoint's default policy mode, so pull that in to its own section. Then, add the docs for EnableDefaultDeny.

While we're at it, clean up obsolete configuration references; we don't support configuration by environment variable anymore.

### CODE REVIEW

File: Documentation/security/policy/intro.rst [hunk 1]
```diff
@@ -14,53 +14,88 @@ Policy Enforcement Modes
...
-
-  This means that endpoints will start without any restrictions and as soon as
-  a rule restricts their ability to receive traffic on ingress or to transmit
-  traffic on egress, then the endpoint goes into whitelisting mode and all
-  traffic must be explicitly allowed.
+  This is the default behavior for policy enforcement. In this mode, endpoints
+  have unrestricted network access until selected by policy. Upon being selected by
+  a policy, the endpoint permits only allowed traffic. This state is per-direction
+  and can be adjusted on a per-policy basis. :ref:`See below<policy_mode_default>` for
+  more details.
```
Comment (qmonnet) [inline Documentation/security/policy/intro.rst]:
```suggestion
  and can be adjusted on a per-policy basis. For more details, :ref:`see the dedicated section on default mode<policy_mode_default>`.
```

File: Documentation/security/policy/intro.rst [hunk 2]
```diff
@@ -14,53 +14,88 @@ Policy Enforcement Modes
...
   enable communications to and from the health endpoint. See :ref:`health_endpoint`.
 
 never
   With never mode, policy enforcement is disabled on all endpoints, even if
   rules do select specific endpoints. In other words, all traffic is allowed
   from any source (on ingress) or destination (on egress).
 
-To configure the policy enforcement mode at runtime for all endpoints managed by a Cilium agent, use:
+To :ref:`configure <k8s_configuration>` the policy enforcement mode, you can adjust the Helm value
+``policyEnforcementMode`` or the corresponding configuration flag ``enable-policy``
```
Comment (qmonnet) [inline Documentation/security/policy/intro.rst]:
```suggestion
To :ref:`configure <k8s_configuration>` the policy enforcement mode, adjust the Helm value
``policyEnforcementMode`` or the corresponding configuration flag ``enable-policy``.
```

File: Documentation/security/policy/intro.rst [hunk 3]
```diff
@@ -14,53 +14,88 @@ Policy Enforcement Modes
...
+.. _policy_mode_default:
 
-If you want to configure the policy enforcement mode at start-time for a particular agent, provide the following flag when launching the Cilium
-daemon:
+Endpoint default policy
+-----------------------
 
-.. code-block:: shell-session
+By default, all egress and ingress traffic is allowed for all endpoints. When
+an endpoint is selected by a network policy, it transitions in to a default-deny
```
Comment (qmonnet) [inline Documentation/security/policy/intro.rst]:
```suggestion
an endpoint is selected by a network policy, it transitions to a default-deny
```

File: Documentation/security/policy/intro.rst [hunk 4]
```diff
@@ -14,53 +14,88 @@ Policy Enforcement Modes
...
+state, where only **explicitly allowed** traffic is permitted. This state is
+per-direction:
 
-    $ cilium-agent --enable-policy={default,always,never} [...]
+* If any rule selects an :ref:`endpoint` and the rule has an ingress
+  section, the endpoint goes into default deny mode for ingress.
```
Comment (qmonnet) [inline Documentation/security/policy/intro.rst]:
```suggestion
  section, the endpoint goes into default-deny mode for ingress.
```

File: Documentation/security/policy/intro.rst [hunk 5]
```diff
@@ -14,53 +14,88 @@ Policy Enforcement Modes
...
+* If any rule selects an :ref:`endpoint` and the rule has an egress section, the
+  endpoint goes into default deny mode for egress.
```
Comment (qmonnet) [inline Documentation/security/policy/intro.rst]:
```suggestion
  endpoint goes into default-deny mode for egress.
```

File: Documentation/security/policy/intro.rst [hunk 6]
```diff
@@ -14,53 +14,88 @@ Policy Enforcement Modes
...
 
-.. code-block:: yaml
+It is possible to create policies that do not enable the default-deny mode for selected
+endpoints. The field ``EnableDefaultDeny`` configures this. Rules with ``EnableDefaultDeny``
+disabled are ignored when determining the default mode.
 
-    - name: CILIUM_ENABLE_POLICY
-      value: always
+For example, this policy causes all DNS traffic to be intercepted, but does not
+block any traffic, even if it is the first policy to apply to an endpoint.
```
Comment (qmonnet) [inline Documentation/security/policy/intro.rst]: What would be a use case for this rule? It could be interesting to mention one in the docs.

**Review by qmonnet — CHANGES_REQUESTED** [review]
Thanks! I've got a few nits, nothing major.

File: Documentation/security/policy/intro.rst (hunk 6 above)
Comment (squeed) [reply Documentation/security/policy/intro.rst]: Good point, added one.

**Review by qmonnet — APPROVED** [review]
Thank you

**Review by christarazi — APPROVED** [review]

**Pull request merged**